### PR TITLE
Make mutable_config_dir configurable + rename

### PR DESCRIPTION
**Pull request by jdavcs**
This replaces #9377 
As per recent discussions on gitter. Ref #9362
I haven't added any new tests or safeguards - it seems to be a straightforward change after all. Unless I'm missing something. I chose managed_config_dir, but I'm open to galaxy_managed_config_dir (or other suggestions).

### CODE REVIEW

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py:137]:
```suggestion
            self.managed_config_dir = config_kwargs.get('managed_config_dir')
            if self.managed_config_dir:
                self.managed_config_dir = os.path.abspath(self.managed_config_dir)
```

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py]: Not sure if you like the helper methods or not, just trying to be consistent, feel free to disregard.

```suggestion
            self.shed_tools_dir = self._in_data_dir('shed_tools')
```

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py]:
```suggestion
                    self.managed_config_dir = self._in_data_dir('config')
```

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py]:
```suggestion
                    self.data_dir = self._in_config_dir('data')
```

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py]:
```suggestion
        self.integrated_tool_panel_config = self._in_managed_config_dir(self.integrated_tool_panel_config)
```

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py]: Question: if the admin sets `managed_config_dir` to the empty string `''`, it is neither resolved to an absolute path (even with my suggestion above, which follows what is done for `config_dir`) nor overridden here.

It's not clear what an admin would intend, but I think we should decide how to interpret that.

The same applies for `config_dir`, `data_dir`.

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py]: Yep, consistent and better.

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py]: I will apply this for now, but I plan to get rid of such cases by setting `path_resolves_to` in the schema - there are quite a few of those, and more..

I don't want to move on with those until #9371 is in - otherwise I'd have to add more changes to it, which are not directly related to its main theme, and I want to avoid that and keep it more or less atomic. It's not an urgent PR, but it's holding up a bunch of modifications I have in mind.. Sorry for being such a pest, but could you take a look at it, please?

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py]: I think setting a config option to `''` is a valid use case. That's how an admin would override the schema default if they wanted no value to be set. It will indeed break things if the `config_dir` path is set to `''`, but so would other unreasonable settings. Besides, in this case galaxy would fail fast and loudly: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L839

So maybe just leave as is?

Otherwise, one way to handle it, I think, would be to change these lines: https://github.com/galaxyproject/galaxy/blob/823526a1fc6c14c868f473f067e68c37ecd5197a/lib/galaxy/config/__init__.py#L140 (6 total) from 
`if self.foo is None:` 
to 
`if not self.foo:`

So, we treat `''` same as "not set", and set it to defaults.
What do you think?

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py:137]: If those options are NOT set by the admin (`managed_config_dir`, `data_dir`, etc.), they are eventually resolved w.r.t. an absolute path. If they ARE set by an admin, I think I would leave it up to the admin. There are many config options that are paths, and if some are made absolute by the system, while others are not, that may be unexpected behavior to an admin, I think? (all paths are equal, but some are more equal than the rest..)

Comment (nsoranzo) [reply lib/galaxy/config/__init__.py:137]: I think you are making an argument for making all paths absolute...

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [reply lib/galaxy/config/__init__.py]: I have to work on training material for the Galaxy Admin Training next week, feel free to have it reviewed by another committer!

File: lib/galaxy/config/__init__.py
Comment (natefoo) [reply lib/galaxy/config/__init__.py:137]: It's ok if the admin sets it relative but is there harm in making it absolute internally?

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [reply lib/galaxy/config/__init__.py]: Changing to `if not` would also be my preferred solution.

Comment (jdavcs) [reply lib/galaxy/config/__init__.py]: OK, can do.

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py]: Thanks! Will do.

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [reply lib/galaxy/config/__init__.py:137]: OK, no quick answer. I'm thinking of (a) ways it may break something and (b) at what point it would be safe to do so. For example, we want to call `strip_deprecated_dir` BEFORE we make user-set path absolute. There may be other issues like this one. Thinking...

Comment (nsoranzo) [reply lib/galaxy/config/__init__.py:137]: Most user-specified paths are made absolute when we call `_in_data_dir()` or sister methods on them, given that `self.data_dir` &c. are already absolute.

Comment (jdavcs) [reply lib/galaxy/config/__init__.py:137]: Exactly. If the user sets `foopath='foo'`, and then it's passed to `_in_data_dir()`, you'll get something like `[abs data_dir path]/foo`. But if we *first* make that an abs path, then `_in_data_dir()` won't affect it and it'll remain `[abs foo path]`. So there will be no way for a user to set a dir to 'foo' and have galaxy place it inside its `data_dir`. In other words, I'm not sure we are not trying to assume too much about user-set paths here.

Comment (nsoranzo) [reply lib/galaxy/config/__init__.py:137]: I'm not advocating for any other changes apart for this one, where the value comes directly from the admin.

Comment (natefoo) [reply lib/galaxy/config/__init__.py:137]: @ic4f does `managed_config_dir` get passed through `_in_data_dir()` later? If so then I suppose it renders @nsoranzo's suggestion moot, unless I misunderstand?

Comment (jdavcs) [reply lib/galaxy/config/__init__.py:137]: it doesn't - I'm currently testing different config setups. So far it appears that if the admin sets it to some value that is a relative str, it stays relative.

Comment (jdavcs) [reply lib/galaxy/config/__init__.py:137]: Double checking, but I think we, indeed, may want to make it absolute - most likely along with `data_dir` and `config_dir`. I'm still checking this..

Comment (jdavcs) [reply lib/galaxy/config/__init__.py:137]: OK, so I'm making the change based on this premise: *if a config option is set by the user to a RELATIVE path, it must be ALWAYS resolved w.r.t some absolute directory by galaxy*.
I'll make the change for `managed_config_dir` here and will address the rest in a separate PR - not to convolute this one. I'll also document this (along with other config-related dev guidelines) - as soon as the rest of pending stuff in config is mostly done.

**Review by nsoranzo — APPROVED** [review]
Fantastic, thanks for bearing with me!

**Comment by jdavcs** [comment]
@nsoranzo well, you were right :-)
Thank you for approving!

**Comment by jdavcs** [comment]
I'm merging this: test failure (py27_unit: `jobs/test_expression_run.py`) is unrelated.

**Pull request merged**
